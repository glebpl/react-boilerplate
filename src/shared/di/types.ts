declare const typeMarker: unique symbol;

export interface Token<T> {
  readonly type: unique symbol;
  // [typeMarker]: T;
}

export type Newable<T> = new () => T;
export type Factory<T> = () => T;
export type Value<T> = T;

export interface CacheItem<T> {
  get(): T;
  getToken(): Token<T>;
  to(object: Newable<T>): CacheItem<T>;
  toFactory(factory: Factory<T>): CacheItem<T>;
  toValue(value: Value<T>): CacheItem<T>;
  setSingletonScope(value: boolean): CacheItem<T>;
}

export type BindFunction = <T = never>(token: Token<T>) => CacheItem<T>;

export interface ResolveFunction {
  <T1 = never>(token1: Token<T1>): T1;
  <T1 = never, T2 = never>(token1: Token<T1>, token2: Token<T2>): [T1, T2];
  <T1 = never, T2 = never, T3 = never>(token1: Token<T1>, token2: Token<T2>, token3: Token<T3>): [T1, T2, T3];
  <T1 = never, T2 = never, T3 = never, T4 = never>(
    token1: Token<T1>,
    token2: Token<T2>,
    token3: Token<T3>,
    token4: Token<T4>
  ): [T1, T2, T3, T4];
  <T1 = never, T2 = never, T3 = never, T4 = never, T5 = never>(
    token1: Token<T1>,
    token2: Token<T2>,
    token3: Token<T3>,
    token4: Token<T4>,
    token5: Token<T5>
  ): [T1, T2, T3, T4, T5];
}
