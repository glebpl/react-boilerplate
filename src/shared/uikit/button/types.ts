import * as React from 'react';

export type Appearance = 'default' | 'danger' | 'link' | 'primary' | 'subtle' | 'warning' | 'accent' | 'accent-subtle';

export type Spacing = 'compact' | 'default';

// Similar to {...A, ...B}
// 1. Remove all overlapping types from First
// 2. Add properties from Second
// https://codesandbox.io/s/native-button-with-nested-elementsclick-bnpjg?file=/src/index.ts
/* This type is intentionally not exported to prevent it from being explicitly referenced in the resulting button types. The alternative would
 * be making this public API and re-exporting from the root */
type Combine<First, Second> = Omit<First, keyof Second> & Second;

export type BaseOwnProps = {
  /** The base styling to apply to the button */
  appearance?: Appearance;
  /** Set the button to autofocus on mount */
  autoFocus?: boolean;
  /** Add a classname to the button */
  className?: string;
  style?: React.CSSProperties; //CSSObject;
  /** Used to 'overlay' something over a button. This is commonly used to display a loading spinner */
  overlay?: React.ReactNode;
  /** Provides a url for buttons being used as a link */
  href?: string;
  /** Places an icon within the button, after the button's text */
  iconAfter?: React.ReactChild;
  /** Places an icon within the button, before the button's text */
  iconBefore?: React.ReactChild;
  /** Set if the button is disabled */
  isDisabled?: boolean;
  /** Change the style to indicate the button is selected */
  isSelected?: boolean;
  isActive?: boolean;
  isHover?: boolean;
  /** Handler to be called on blur */
  onBlur?: React.FocusEventHandler<HTMLElement>;
  /** Handler to be called on click. The second argument can be used to track analytics data. See the tutorial in the analytics-next package for details */
  onClick?: (e: React.MouseEvent<HTMLElement>) => void;
  /** Handler to be called on focus */
  onFocus?: React.FocusEventHandler<HTMLElement>;
  /** Set the amount of padding in the button */
  spacing?: Spacing;
  /** Pass target down to a link within the button component, if a href is provided */
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>['target'];
  /** Pass type down to a button */
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>['type'];
  /** Option to fit button width to its parent width */
  shouldFitContainer?: boolean;
  /** Text content to be rendered in the button */
  children?: React.ReactNode;
  /** A `testId` prop is provided for specified elements, which is a unique string that appears as a data attribute `data-testid` in the rendered code, serving as a hook for automated tests */
  testId?: string;

  /* An optional way of changing what button renders
    - `React.ElementType`: force the button to render whatever html element you want (eg "div")
    - `React.ComponentType<React.AllHTMLAttributes<HTMLElement>>`: pass in a component that can accept any HTMLAttribute as a prop and render whatever you would like to
  */
  component?: React.ComponentType<React.AllHTMLAttributes<HTMLElement>> | React.ElementType;

  /** Additional information to be included in the `context` of analytics events that come from button */
  analyticsContext?: Record<string, any>;
  ref?: React.Ref<HTMLElement>;
};

export type BaseProps = Combine<
  // Not allowing 'disabled' as a prop, as we control disabled through isDisabled
  Omit<React.AllHTMLAttributes<HTMLElement>, 'disabled'>,
  BaseOwnProps
>;

export interface SplitButtonGroupProps {
  appearance?: Appearance;
  /** Set if all items in group are disabled */
  isDisabled?: boolean;
  /** Set the amount of padding in group items */
  spacing?: Spacing;
}

export interface SplitButtonGroupItemProps extends Required<SplitButtonGroupProps> {
  style: React.CSSProperties;
}
