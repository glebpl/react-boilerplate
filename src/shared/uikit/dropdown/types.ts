import { ReactNode, KeyboardEvent, MouseEvent, HTMLAttributes, SyntheticEvent } from 'react';
import { ButtonProps } from '../button';
import { ContentProps as PopupContentProps, PopupProps, TriggerProps } from '../popup';
import { EmotionThemeProps } from '../theme';

export type CloseCallback = (event?: KeyboardEvent | MouseEvent | SyntheticEvent, returnFocusToRef?: boolean) => void;

export type DropdownItemClickCallback = (e: MouseEvent | KeyboardEvent, shouldCloseDropdown: boolean) => void;

export interface DropdownItemProps extends EmotionThemeProps {
  /**
   * Primary content for the item.
   */
  children: ReactNode;

  /**
   * Makes the element appear disabled as well as removing interactivity.
   */
  isDisabled?: boolean;

  /**
   * Hides the item.
   */
  isHidden?: boolean;

  /**
   * Should dropdown close when item clicked
   */
  shouldCloseDropdown?: boolean;

  /**
   * Event that is triggered when the element is clicked.
   */
  onClick?: (e: MouseEvent | KeyboardEvent) => void;

  onKeyDown?: (e: KeyboardEvent) => void;

  /**
   * Link to another page.
   */
  href?: string;

  /**
   * Where to display the linked URL,
   * see [anchor information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) on mdn for more information.
   */
  target?: string;

  /**
   * Adds a title attribute to the root item element.
   */
  title?: string;

  /**
   * Is not applied. Requires Tooltip review
   */
  // tooltip?: ReactNode;

  /**
   * Element to render before the item text.
   * Generally should be an icon component..
   */
  elemBefore?: ReactNode;

  /**
   * Element to render after the item text.
   * Generally should be an icon component.
   */
  elemAfter?: ReactNode;
  testId?: string;
}

export interface SelectableDropdownItemProps extends DropdownItemProps {
  id: string;
  isSelected?: boolean;
}

export interface DropdownItemGroupProps extends Omit<HTMLAttributes<HTMLDivElement>, 'title'> {
  elemAfter?: ReactNode;
  title?: ReactNode;
}

export interface DropdownTriggerProps<T extends HTMLElement = HTMLElement> extends TriggerProps<T> {
  isDisabled: boolean;
  isSelected: boolean;
  onClick: (e: MouseEvent | KeyboardEvent) => void;
  onKeyDown: (e: KeyboardEvent) => void;
  testId?: string;
}

export type DrodownTriggerRenderFn = (props: Partial<DropdownTriggerProps>) => ReactNode;

export interface DropdownContentProps extends PopupContentProps {}

export interface DropdownProps extends EmotionThemeProps {
  /**
   * By default dropdown will be closed also when reference element becomes invisible
   * If closeOnTriggerEscape === false dropdown stays visible until it escapes visible area even if reference element is already invisible.
   * It may be useful in modals and in scrollable content.
   * Default is true
   */
  closeOnTriggerEscape?: boolean;
  /**
   * HTML props passed to dropdown content element
   */
  contentProps?: HTMLAttributes<HTMLDivElement>;
  /**
   * Placements to be tried if dropdown can not be shown using placement
   * See https://popper.js.org/docs/v2/modifiers/flip/
   */
  fallbackPlacements?: PopupProps['fallbackPlacements'];
  isDisabled?: boolean;
  isOpen?: boolean;
  defaultIsOpen?: boolean;

  /**
   *  onOpen and onClose called when a state change request is received
   */
  onOpen?: (event?: KeyboardEvent | MouseEvent) => void;
  onClose?: (event?: KeyboardEvent | MouseEvent | SyntheticEvent) => void;

  /**
   * Called when dropdown is open (visible) and correctly positioned
   */
  afterOpen?: () => void;
  placement?: PopupProps['placement'];
  tooltip?: ReactNode;
  /**
   * Element which will trigger the dropdown menu to open and close.
   * If string is passed default button with chevron will be rendered
   */
  trigger: ReactNode | DrodownTriggerRenderFn;
  /** Props to pass through to the trigger button. See @atlaskit/button for allowed props. */
  triggerButtonProps?: ButtonProps;
  /**
   As dropdown is composed of different components, we pass down the testId to the sub component you want to test:
   - `testId--trigger` - trigger.
   - `testId--content` - content.
   - `testId--popup` - popup container.
   */
  testId?: string;
  zIndex?: number;

  contentRef?: (node: HTMLDivElement) => void;
  popupRef?: (node: HTMLDivElement) => void;
}

export interface DropdownEventContextAPI {
  onItemClick: DropdownItemClickCallback;
  onItemKeyDown: (e: KeyboardEvent) => void;
  onContentKeyDown: (e: KeyboardEvent) => void;
}
