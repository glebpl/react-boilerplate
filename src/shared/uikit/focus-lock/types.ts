import type { DetailedHTMLProps, HTMLAttributes } from 'react';

type DivProps = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;

export interface FocusLockProps extends DivProps {
  /**
   * If true - focus will be set to first tabbable element, if no enabled tabbable elements inside, container will be focused.
   * Safari settings are not respected, so if first tabbable is link or button it will be focused.
   * If false - focus will be set to parent container added by FocusLock component. So next Tab will move focus to first tabbable
   * If html element - focus it
   * If function - focus will be set to element returned by it
   * If undefined - noop
   * Default is undefined
   */
  autoFocus?: boolean | HTMLElement | null | (() => HTMLElement | null);
  /**
   * Allows usage as styled(FocusLock)
   */
  className?: string;
  /**
   * Whether the focus lock is active or not.
   * Default is true
   */
  isEnabled?: boolean;
  /**
   * Should focus be returned to element which was active before FocusLock mounting
   * Default is true
   */
  shouldReturnFocus?: boolean;
  testId?: string;
}

export interface MoveFocusInsideProps extends DivProps {
  /**
   * Allows usage as styled(MoveFocusInside)
   */
  className?: string;
  /**
   * Default is true
   */
  isEnabled?: boolean;
}
