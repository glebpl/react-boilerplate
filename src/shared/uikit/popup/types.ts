import {
  AriaAttributes,
  Dispatch,
  KeyboardEvent,
  MouseEvent,
  ReactNode,
  Ref,
  RefCallback,
  RefObject,
  SetStateAction
} from 'react';
import { ButtonProps } from '../button';
import { BasePlacement, Placement, PopperChildrenProps, PopperProps, VirtualElement } from '../popper';
import { FocusLockProps } from '../focus-lock';

export type PopupAppearance = 'default' | 'accent';

export interface TriggerProps<T extends HTMLElement = HTMLElement>
  extends Pick<AriaAttributes, 'aria-expanded' | 'aria-haspopup'> {
  ref: Ref<T>;
}

export type PopupTriggerRenderFn = (props: TriggerProps) => ReactNode;

export type PopupRef = HTMLDivElement | null;

export type TriggerRef = HTMLElement | null;

export interface ContentProps {
  /**
   * Will reposition the popup if any of the content has changed.
   * Useful for when positions change and the popup was not aware.
   */
  update: PopperChildrenProps['update'];

  /**
   * Passed through from the parent popup.
   */
  isOpen: boolean;

  /**
   * Received from parent popup.
   */
  isReferenceHidden: boolean | undefined;

  /**
   * Passed through from the parent popup.
   */
  onClose: (() => void) | undefined;

  /**
   * Escape hatch to set the initial focus for a specific element when the popup is opened.
   */
  setInitialFocusRef: Dispatch<SetStateAction<HTMLElement | null>>;
}

export interface PopupComponentProps extends FocusLockProps {
  /**
   * Placement passed through by the parent popup.
   */
  'data-placement'?: Placement;
}

export interface PopupProps {
  appearance?: PopupAppearance;
  /**
   * Used to either show or hide the popup.
   * When set to `false` popup will not render anything to the DOM.
   */
  isOpen: boolean;

  /**
   * Render props used to anchor the popup to your content.
   * Make this an interactive element,
   * such as an @atlaskit/button component.
   */
  trigger?: PopupTriggerRenderFn;

  /**
   * Props passed to popup container component
   */
  componentProps?: PopupComponentProps;

  /**
   * Render props for content that is displayed inside the popup.
   */
  content: (props: ContentProps) => React.ReactNode;

  /**
   * The boundary element that the popup will check for overflow.
   * Defaults to `"clippingParents"` which are parent scroll containers,
   * but can be set to any element.
   */
  boundary?: 'clippingParents' | HTMLElement;

  /**
   * Optional real or virtual reference. May be used instead of trigger to show stationary popups
   * or popups for buttons opening another popups/dropdowns
   */
  referenceElement?: PopperProps<any>['referenceElement'];

  /**
   * The root boundary that the popup will check for overflow.
   * Defaults to `"viewport"` but can be set to `"document"`.
   */
  rootBoundary?: 'viewport' | 'document';

  /**
   * Id that is assigned to the popup container element.
   */
  id?: string;

  onPopupEscaped?: () => void;
  onReferenceEscaped?: () => void;

  /**
   * Distance the popup should be offset from the reference in the format of [along, away] (units in px).
   * Defaults to [0, 8] - which means the popup will be 8px away from the edge of the reference specified
   * by the `placement` prop.
   */
  offset?: [number, number];

  /**
   * Placement of where the popup should be displayed relative to the trigger element.
   * Defaults to `"auto"`.
   */
  placement?: Placement;
  popupRef?: RefCallback<HTMLDivElement> | RefObject<HTMLDivElement> | null;

  /**
   * Allows the Popup to be placed on the opposite side of its trigger if it does not fit in the viewport.
   * Defaults to `true`.
   */
  shouldFlip?: boolean;

  /**
   * Placements which may be used to flip
   */
  fallbackPlacements?: Placement[];

  /**
   * A `testId` prop is provided for specified elements,
   * which is a unique string that appears as a data attribute `data-testid` in the rendered code,
   * serving as a hook for automated tests.
   */
  testId?: string;

  /**
   * Handler that is called when the popup wants to close itself.
   * Generally this will be either when clicking away from the popup or pressing the escape key.
   * You'll want to use this to set open state accordingly and then pump it back into the `isOpen` prop.
   */
  onClose?: (event?: React.SyntheticEvent) => void;

  /**
   * Called after popup opened
   */
  onOpen?: () => void;

  afterOpen?: () => void;

  /**
   * Z-index that the popup should be displayed in.
   * This is passed to the portal component.
   * Defaults to `layers.layer()` from `@atlaskit/theme`.
   */
  zIndex?: number;

  /**
   * Controls whether the popup takes focus when opening.
   * Defaults to `true`.
   */
  autoFocus?: boolean;
  /**
   * Will Popup use focus locking inside it's content or leave it's managing to parent component
   * Default is "true"
   */
  lockFocus?: boolean;
  /**
   * Will focus be returned to trigger when popup close
   * Default is "false" (managed by useCloseManager)
   */
  shouldReturnFocusToTrigger?: boolean;
}

export interface AccentPopupAction extends ButtonProps {
  text: ReactNode;
}

export interface AccentPopupProps extends Omit<PopupProps, 'content'> {
  action?: AccentPopupAction;
  actions?: AccentPopupAction[];
  content: ReactNode;
  // accent popup may be placed only bottom/top/left/right due to arrow
  fallbackPlacements?: BasePlacement[];
  placement?: BasePlacement;
  title?: ReactNode;
}

export interface AccentedProps {
  accentActions?: AccentPopupAction[];
  accentContent: ReactNode;
  accentTitle?: ReactNode;
  accentedElement: HTMLElement | VirtualElement | null;
  lockFocus?: boolean;
}

export interface AccentManager {
  close: () => void;
  isOpen: () => boolean;
}

export type CloseManagerHook = Pick<PopupProps, 'isOpen' | 'onClose'> & {
  manageFocusReturn: boolean;
  popupRef: PopupRef;
  triggerRef: TriggerRef;
};

export type RepositionOnUpdateProps = {
  content: (props: ContentProps) => React.ReactNode;
  update: PopperChildrenProps['update'];
};
