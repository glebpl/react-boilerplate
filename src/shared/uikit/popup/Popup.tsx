/** @jsx jsx */
// noinspection ES6UnusedImports
import { jsx, SerializedStyles } from '@emotion/react';
import * as React from 'react';
import { FC, memo, useCallback, useEffect, useMemo, useState } from 'react';
import { Modifier } from 'react-popper';
import { FocusLock } from '../focus-lock';
import { FlipModifierOptions, Manager, Popper, Reference } from '../popper';
import Portal from '../portal';
import { EmotionTheme, layers } from '../theme';
import setRef from '../utils/setRef';
import { RepositionOnUpdate } from './RepositionOnUpdate';
import { PopupProps } from './types';
import { useCloseManager } from './useCloseManager';
import { createPopupCss } from './styles';

export const Popup: FC<PopupProps> = memo(
  ({
    appearance = 'default',
    componentProps = {},
    content,
    isOpen,
    id,
    offset,
    testId,
    trigger,
    onClose,
    onOpen,
    afterOpen,
    boundary,
    rootBoundary = 'viewport',
    placement = 'auto',
    shouldFlip = true,
    fallbackPlacements,
    zIndex = layers.layer(),
    onPopupEscaped,
    onReferenceEscaped,
    lockFocus = true,
    referenceElement,
    shouldReturnFocusToTrigger = false,
    ...rest
  }: PopupProps) => {
    // pass undefined explicitly to override default 'true' value
    const autoFocus = 'autoFocus' in rest ? rest.autoFocus : true;
    const [popupRef, setPopupRef] = useState<HTMLDivElement | null>(null);
    const [triggerRef, setTriggerRef] = useState<HTMLElement | null>(null);
    const [initialFocusRef, setInitialFocusRef] = useState<HTMLElement | null>(null);
    const modifiers = useMemo(() => {
      const options: Partial<FlipModifierOptions> = {
        rootBoundary,
        boundary
      };
      if (fallbackPlacements) {
        options.fallbackPlacements = fallbackPlacements;
      }
      const cm: Modifier<'flip' | 'afterMain' | 'afterWrite'>[] = [
        {
          name: 'flip',
          enabled: shouldFlip,
          options
        }
      ];
      if (onPopupEscaped || onReferenceEscaped) {
        const afterMainModifier: Modifier<'afterMain'> = {
          name: 'afterMain',
          phase: 'afterMain',
          enabled: true,
          requires: ['hide'],
          fn: ({ state }): void => {
            const { hide: hideModifierData } = state.modifiersData;
            if (hideModifierData) {
              hideModifierData.isReferenceHidden && onReferenceEscaped?.();
              hideModifierData.hasPopperEscaped && onPopupEscaped?.();
            }
          }
        };

        cm.push(afterMainModifier);
      }
      if (afterOpen) {
        const afterWriteModifier: Modifier<'afterWrite'> = {
          name: 'afterWrite',
          phase: 'afterWrite',
          enabled: true,
          requires: [],
          fn: (): void => {
            afterOpen?.();
          }
        };
        cm.push(afterWriteModifier);
      }
      return cm;
    }, [fallbackPlacements, shouldFlip, rootBoundary, boundary, onPopupEscaped, onReferenceEscaped, afterOpen]);

    useCloseManager({ isOpen, manageFocusReturn: !shouldReturnFocusToTrigger, onClose, popupRef, triggerRef });

    useEffect(() => {
      isOpen && onOpen && onOpen();
    }, [isOpen, onOpen]);

    const popupCss = useCallback(
      (theme: EmotionTheme): SerializedStyles => createPopupCss(theme, appearance),
      [appearance]
    );

    return (
      <Manager>
        {referenceElement ? null : trigger ? (
          <Reference>
            {({ ref }) =>
              trigger({
                ref: (node: HTMLElement | null) => {
                  if (node) {
                    if (typeof ref === 'function') {
                      ref(node);
                    } else {
                      (ref as React.MutableRefObject<HTMLElement>).current = node;
                    }

                    setTriggerRef(node);
                  }
                },
                'aria-expanded': isOpen,
                'aria-haspopup': true
              })
            }
          </Reference>
        ) : null}
        {isOpen && (
          <Portal zIndex={zIndex}>
            <Popper placement={placement} offset={offset} modifiers={modifiers} referenceElement={referenceElement}>
              {({ ref, style: popperCSSProps, placement: popupPlacement, update, isReferenceHidden }) => (
                <FocusLock
                  {...componentProps}
                  css={popupCss}
                  autoFocus={initialFocusRef || autoFocus}
                  isEnabled={lockFocus}
                  shouldReturnFocus={shouldReturnFocusToTrigger}
                  data-placement={popupPlacement}
                  id={id}
                  ref={(node: HTMLDivElement) => {
                    setRef(ref, node);
                    setPopupRef(node);
                    if (rest.popupRef) {
                      setRef(rest.popupRef, node);
                    }
                  }}
                  style={popperCSSProps}
                  testId={testId}
                >
                  <RepositionOnUpdate content={content} update={update}>
                    {content({
                      update,
                      isOpen,
                      onClose,
                      setInitialFocusRef,
                      isReferenceHidden
                    })}
                  </RepositionOnUpdate>
                </FocusLock>
              )}
            </Popper>
          </Portal>
        )}
      </Manager>
    );
  }
);

export default Popup;
