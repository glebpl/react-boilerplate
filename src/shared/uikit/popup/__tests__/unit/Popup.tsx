import { fireEvent, render, waitFor } from '@testing-library/react';
import { mount } from 'enzyme';
import * as React from 'react';
import { useRef } from 'react';
import * as raf from '../../../utils/_testutils/raf';
import Popup from '../../Popup';
import { ContentProps, PopupProps, TriggerProps } from '../../types';

// override requestAnimationFrame letting us execute it when we need
raf.replace();

type ButtonRef = React.Ref<HTMLButtonElement>;

const defaultProps: PopupProps = {
  content: () => <div>content</div>,
  isOpen: false,
  trigger: (props: TriggerProps) => (
    <button {...props} ref={props.ref as ButtonRef}>
      trigger
    </button>
  )
};

it('renders the trigger correctly when the popup is not open', () => {
  const trigger = (props: TriggerProps) => (
    <button {...props} ref={props.ref as ButtonRef}>
      trigger
    </button>
  );

  const { getByText } = render(<Popup {...defaultProps} isOpen={false} trigger={trigger} />);
  const triggerEl = getByText('trigger');

  expect({
    'aria-expanded': triggerEl.getAttribute('aria-expanded'),
    'aria-haspopup': triggerEl.getAttribute('aria-haspopup')
  }).toEqual({
    'aria-expanded': 'false',
    'aria-haspopup': 'true'
  });
});

it('renders the trigger correctly when the popup is open', async () => {
  const trigger = (props: TriggerProps) => (
    <button {...props} ref={props.ref as ButtonRef}>
      trigger
    </button>
  );
  const { getByText } = render(<Popup {...defaultProps} isOpen trigger={trigger} />);

  const triggerEl = getByText('trigger');

  await waitFor(() => {
    expect({
      'aria-expanded': triggerEl.getAttribute('aria-expanded'),
      'aria-haspopup': triggerEl.getAttribute('aria-haspopup')
    }).toEqual({
      'aria-expanded': 'true',
      'aria-haspopup': 'true'
    });
  });
});

it('does not render the content when the popup is not open', () => {
  const { queryByText } = render(<Popup {...defaultProps} />);
  expect(queryByText('content')).not.toBeInTheDocument();
});

it('renders the content correctly when the popup is open', async () => {
  const { getByText } = render(<Popup {...defaultProps} isOpen />);
  await waitFor(() => {
    expect(getByText('content')).toBeInTheDocument();
  });
});

it('renders the content correctly when the popup is opened', async () => {
  const { getByText, rerender } = render(<Popup {...defaultProps} />);
  rerender(<Popup {...defaultProps} isOpen />);
  await waitFor(() => {
    expect(getByText('content')).toBeInTheDocument();
  });
});

it('does not call onClose after pressing escape when the popup is not open', () => {
  const onClose = jest.fn();
  const { baseElement } = render(<Popup {...defaultProps} onClose={onClose} />);

  fireEvent.keyDown(baseElement, { key: 'Escape', code: 'Escape' });

  expect(onClose).not.toHaveBeenCalled();
});

it('calls onClose after pressing escape when the popup is open', async () => {
  const onClose = jest.fn();
  const { baseElement } = render(<Popup {...defaultProps} isOpen onClose={onClose} />);

  fireEvent.keyDown(baseElement, { key: 'Escape', code: 'Escape' });

  await waitFor(() => {
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

it('calls onClose after pressing escape when the popup is opened', async () => {
  const onClose = jest.fn();
  const { baseElement, rerender } = render(<Popup {...defaultProps} onClose={onClose} />);

  rerender(<Popup {...defaultProps} isOpen onClose={onClose} />);

  fireEvent.keyDown(baseElement, { key: 'Escape', code: 'Escape' });

  await waitFor(() => {
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

it('calls onClose after clicking on the trigger when the popup is open', () => {
  const onClose = jest.fn();
  const { getByText } = render(
    <Popup {...defaultProps} isOpen onClose={onClose} trigger={() => <button>trigger</button>} />
  );

  fireEvent.click(getByText('trigger'));

  expect(onClose).toHaveBeenCalledTimes(1);
});

it('calls onClose after clicking on the trigger when the popup is opened', () => {
  const onClose = jest.fn();
  const trigger = () => <button>trigger</button>;
  const { getByText, rerender } = render(<Popup {...defaultProps} onClose={onClose} trigger={trigger} />);

  rerender(<Popup {...defaultProps} isOpen onClose={onClose} trigger={trigger} />);

  fireEvent.click(getByText('trigger'));

  expect(onClose).toHaveBeenCalledTimes(1);
});

it('calls onClose after clicking outside of the popup when the popup is open', async () => {
  const onClose = jest.fn();
  const { baseElement } = render(<Popup {...defaultProps} isOpen onClose={onClose} />);

  fireEvent.click(baseElement);

  await waitFor(() => {
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

it('calls onClose after clicking outside of the popup when the popup is opened', async () => {
  const onClose = jest.fn();
  const { baseElement, rerender } = render(<Popup {...defaultProps} onClose={onClose} />);
  rerender(<Popup {...defaultProps} isOpen onClose={onClose} />);

  fireEvent.click(baseElement);

  await waitFor(() => {
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

it('calls onClose when calling onClose within the content', async () => {
  const onClose = jest.fn();
  const { getByText } = render(
    <Popup
      {...defaultProps}
      content={({ onClose: onClick }) => <button onClick={onClick}>x</button>}
      isOpen
      onClose={onClose}
    />
  );

  fireEvent.click(getByText('x'));

  await waitFor(() => {
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

it('does not focus the content when the popup is open', async () => {
  const { getByText } = render(<Popup {...defaultProps} isOpen />);
  await waitFor(() => {
    expect(getByText('content')).not.toHaveFocus();
  });
});

it('does not focus the content when the popup is opened', async () => {
  const { getByText, rerender } = render(<Popup {...defaultProps} />);
  rerender(<Popup {...defaultProps} isOpen />);
  await waitFor(() => {
    expect(getByText('content')).not.toHaveFocus();
  });
});

it('focuses the first focusable element inside of the content when the popup is opened', async () => {
  const { getByText } = render(
    <Popup
      trigger={(props: TriggerProps) => (
        <button {...props} ref={props.ref as ButtonRef}>
          trigger
        </button>
      )}
      content={({ setInitialFocusRef }) => <button ref={setInitialFocusRef}>focused content</button>}
      isOpen
    />
  );

  raf.step();
  raf.step();

  await waitFor(() => {
    expect(getByText('focused content')).toHaveFocus();
  });
});

it('focuses the specified element inside of the content when the popup is opened', async () => {
  const content = ({ setInitialFocusRef }: ContentProps) => <button ref={setInitialFocusRef}>focused content</button>;

  const { getByText, rerender } = render(<Popup {...defaultProps} content={content} />);

  rerender(<Popup {...defaultProps} content={content} isOpen />);

  raf.step();
  raf.step();

  await waitFor(() => {
    expect(getByText('focused content')).toHaveFocus();
  });
});

it('popup stays open if propagation is stopped on an event before it reaches window', async () => {
  const content = () => (
    <button
      onClick={event => {
        event.stopPropagation();
      }}
    >
      Popup content
    </button>
  );

  const { findAllByText } = render(
    <div>
      <Popup {...defaultProps} content={content} isOpen />
      <Popup {...defaultProps} content={content} isOpen />
    </div>
  );

  fireEvent.click((await findAllByText('Popup content'))[0]);

  expect((await findAllByText('Popup content'))[1]).toBeDefined();
});

it('popup stays open when clicked element (which is inside content) is removed from the DOM', async () => {
  const onClose = jest.fn();
  const Component = () => {
    const ref = useRef<HTMLButtonElement | null>(null);

    return (
      <button ref={ref} onClick={() => ref.current && ref.current.remove()}>
        Button content
      </button>
    );
  };

  const { getByText } = render(<Popup {...defaultProps} onClose={onClose} content={() => <Component />} isOpen />);

  fireEvent.click(getByText('Button content'));

  await waitFor(() => {
    expect(onClose).toHaveBeenCalledTimes(0);
  });
});

it('matches snapshot', async () => {
  const popup = mount(<Popup {...defaultProps} isOpen />);
  await waitFor(() => {
    expect(popup).toMatchSnapshot();
  });
});
