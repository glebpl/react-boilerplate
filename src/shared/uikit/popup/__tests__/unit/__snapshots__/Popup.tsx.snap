// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
.emotion-0 {
  background-color: #FFFFFF;
  border-radius: 3px;
  color: inherit;
  box-shadow: 0 3px 5px rgba(9, 30, 66, 0.2),0 0 1px rgba(9, 30, 66, 0.31);
  box-sizing: border-box;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  overflow: auto;
  z-index: 400;
}

.emotion-0:focus {
  outline: none;
}

<Memo()
  content={[Function]}
  isOpen={true}
  trigger={[Function]}
>
  <Manager>
    <Reference>
      <button
        aria-expanded={true}
        aria-haspopup={true}
      >
        trigger
      </button>
    </Reference>
    <Portal
      zIndex={400}
    >
      <Portal
        containerInfo={
          .emotion-0 {
  background-color: #FFFFFF;
  border-radius: 3px;
  color: inherit;
  box-shadow: 0 3px 5px rgba(9, 30, 66, 0.2),0 0 1px rgba(9, 30, 66, 0.31);
  box-sizing: border-box;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  overflow: auto;
  z-index: 400;
}

.emotion-0:focus {
  outline: none;
}

<div
            class="atlaskit-portal"
            style="z-index: 400;"
          >
            <div
              class="emotion-0"
              data-focus-lock="true"
              data-placement="auto"
              style="position: fixed; left: 0px; top: 0px;"
              tabindex="-1"
            >
              <div>
                content
              </div>
            </div>
          </div>
        }
      >
        <Popper
          modifiers={
            Array [
              Object {
                "enabled": true,
                "name": "flip",
                "options": Object {
                  "boundary": undefined,
                  "rootBoundary": "viewport",
                },
              },
            ]
          }
          placement="auto"
        >
          <Popper
            modifiers={
              Array [
                Object {
                  "name": "flip",
                  "options": Object {
                    "boundary": "clippingParents",
                    "flipVariations": false,
                    "padding": 5,
                    "rootBoundary": "viewport",
                  },
                },
                Object {
                  "name": "preventOverflow",
                  "options": Object {
                    "padding": 5,
                    "rootBoundary": "viewport",
                  },
                },
                Object {
                  "name": "offset",
                  "options": Object {
                    "offset": Array [
                      0,
                      8,
                    ],
                  },
                },
                Object {
                  "enabled": true,
                  "name": "flip",
                  "options": Object {
                    "boundary": undefined,
                    "rootBoundary": "viewport",
                  },
                },
              ]
            }
            placement="auto"
            strategy="fixed"
          >
            <ForwardRef(FocusLockComponent)
              autoFocus={true}
              className="emotion-0"
              data-placement="auto"
              isEnabled={true}
              shouldReturnFocus={false}
              style={
                Object {
                  "left": "0",
                  "position": "fixed",
                  "top": "0",
                }
              }
            >
              <div
                className="emotion-0"
                data-focus-lock="true"
                data-placement="auto"
                onKeyDown={[Function]}
                style={
                  Object {
                    "left": "0",
                    "position": "fixed",
                    "top": "0",
                  }
                }
                tabIndex={-1}
              >
                <RepositionOnUpdate
                  content={[Function]}
                  update={[Function]}
                >
                  <div>
                    content
                  </div>
                </RepositionOnUpdate>
              </div>
            </ForwardRef(FocusLockComponent)>
          </Popper>
        </Popper>
      </Portal>
    </Portal>
  </Manager>
</Memo()>
`;
